Add support for a suspendOnRestore option in JDWP

mikezhang1234567890 opened the pull request:
CRIU supports Java debugger during checkpoint and restore

Added support for a `VM_RESTORE` `JVMTI` event;
Added command line option `suspendOnRestore=[y|n]` to determine if the VM restored from a checkpoint is to be suspended;
`-agentlib:jdwp=transport=dt_socket,server=y,suspend=[y|n]` assumes `suspendOnRestore=y`;
Always respect `suspendOnRestore=n` regardless the order of `suspend` and `suspendOnRestore`, both
`-agentlib:jdwp=transport=dt_socket,server=y,suspendOnRestore=n,suspend=y` and `-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,suspendOnRestore=n` don't suspend the JVM post-restore.

~depends on https://github.com/eclipse-openj9/openj9/pull/19405~ merged

Signed-off-by: Mike Zhang <mike.h.zhang@ibm.com>
Signed-off-by: Jason Feng <fengj@ca.ibm.com>

tajila commented:
@JasonFengJ9 Please review

JasonFengJ9 commented:
> 15:10:43  src/jdk.jdi/share/classes/com/sun/jdi/event/VMRestoreEvent.java
15:10:43  	IBM Portions Copyright should be used in this file

This is a new file, it can go to closed/src, and the Copyright is
```
/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
/*
 * ===========================================================================
 * (c) Copyright IBM Corp. 2023, 2023 All Rights Reserved
 * ===========================================================================
 */
```

In addition, there is a line ending check error
```
11:06:26  src/java.se/share/data/jdwp/jdwp.spec:30: trailing whitespace.
```

mikezhang1234567890 commented:
> Any reason it can't be in the openj9 repo instead?

Not sure if there's any reason it can't be in the OpenJ9 repo instead, but I put it in the extension repo since the `jdk.jdi` package needed changes to support the new VM restore event anyways. 

I'll see if moving it to closed/src or the OpenJ9 repo makes more sense.

JasonFengJ9 reviewed (commented):
This is a CRIU feature, the java code can be decorated w/ `CRIU_SUPPORT`, and the native code w/ `J9VM_OPT_CRIU_SUPPORT`, just created
* https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/592

mikezhang1234567890 commented:
Rebased on top of https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/592/commits/d1ba4a114d70c6b88554dc9ec30c510d0b0622df and added ifdefs so that the new code only compiles with CRIU enabled.

Also addressed formatting review comments.

@JasonFengJ9 ready for another look.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1252,6 +1258,46 @@ cbVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+static void JNICALL
+cbVMRestore(jvmtiEnv *jvmti_env, ...)
+{
+    EventInfo info;
```
```suggestion
    EventInfo info = {0};
```

keithc-ca replied:
Please *don't* do this; there's already a use of `memset()` below.

mikezhang1234567890 replied:
Reverted this change in https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/1396459ab0d1f728e71ed55dd10af33c53c2c8b7 so that the object is just declared, since it will be memset to 0 anyways.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1252,6 +1258,46 @@ cbVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+    JNIEnv *env = NULL;
+    jthread thread = NULL;
+    va_list ap;
+    jbyte suspendPolicy;
```
```suggestion
    jbyte suspendPolicy = 0;
```

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -605,6 +623,35 @@ handleReportVMInitCommand(JNIEnv* env, ReportVMInitCommand *command)
+static void
+handleReportVMRestoreCommand(JNIEnv* env, ReportVMRestoreCommand *command)
+{
+    PacketOutputStream out;
```
```suggestion
    PacketOutputStream out = {0};
```

mikezhang1234567890 replied:
Same as above, reverted change in https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/1396459ab0d1f728e71ed55dd10af33c53c2c8b7 since it will get memset.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -1167,6 +1219,23 @@ eventHelper_reportVMInit(JNIEnv *env, jbyte sessionID, jthread thread, jbyte sus
+    if (NULL == command) {
+        EXIT_ERROR(AGENT_ERROR_OUT_OF_MEMORY,"HelperCommmand");
+    }
+    (void)memset(command, 0, sizeof(*command));
```
The cast is not required.

JasonFengJ9 reviewed (commented):
* https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/592
is merged, please rebase.

mikezhang1234567890 commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1918,9 +1923,42 @@ map2jvmtiError(jdwpError error)
 }
 
+static jvmtiExtensionEventInfo*
+find_ext_event(jvmtiEnv *jvmti, const char *ename)
```
Added ifdefs in https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/739e564e4e2760e017d6c47440963836a00c2cc9

mikezhang1234567890 commented:
Rebased on the merged commit and updated to address review comments.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -74,6 +79,9 @@ static jboolean initOnUncaught = JNI_FALSE; /* init when uncaught exc thrown */
 
 static char *launchOnInit = NULL;           /* launch this app during init */
 static jboolean suspendOnInit = JNI_TRUE;   /* suspend all app threads after init */
+#if defined(J9VM_OPT_CRIU_SUPPORT)
```
This requires `#include "j9cfg.h"` which means a dependency on https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/596.

I see that include in `debugInit.h`: why is `debugInit.h` not included here?

JasonFengJ9 replied:
This PR was based on an earlier PR
* https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/592
https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/596 is an improvement though backporting of this PR depends on both PRs.

Note: this file includes `util.h` which includes `debugInit.h` indeed.

keithc-ca replied:
I think every file that uses any flag from `j9cfg.h` should include that header file directly, otherwise we will silently omit code guarded with `#if defined(J9VM_OPT_CRIU_SUPPORT)`, for example, leading to inconsistencies.

JasonFengJ9 replied:
Do we want to include `j9cfg.h` in both `debugInit.c` and `debugInit.h`, similar comments for other c source/header files?

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -396,6 +423,30 @@ cbEarlyVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
 }
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static void JNICALL
```
I'm not sure `static` and `JNICALL` can be used together.

mikezhang1234567890 replied:
The early callback functions here and normal callback functions in `eventHandler.c` all use this signature.

keithc-ca commented on src/java.se/share/data/jdwp/jdwp.spec:
```diff
@@ -3140,6 +3145,17 @@ JDWP "Java(tm) Debug Wire Protocol"
+                      (int requestID "Request that generated event")
+                        (threadObject thread "The thread restoring the VM from a checkpoint.")
+                    )
+/*[ENDIF] CRIU_SUPPORT*/
```
Please add a space before `*/` just like you would in C/C++ code.

keithc-ca commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTY.java:
```diff
@@ -99,6 +104,14 @@ public void vmStartEvent(VMStartEvent se)  {
 
+/*[IF CRIU_SUPPORT]*/
+    @Override
+    public void vmRestoreEvent(VMRestoreEvent se)  {
```
The name `se` might be appropriate for `VMStartEvent` above, but it seems wrong here.

keithc-ca replied:
How about `re`?

mikezhang1234567890 replied:
Should already be changed to use `re`.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -396,6 +423,30 @@ cbEarlyVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+
+    LOG_CB(("cbEarlyVMRestore"));
+    if (gdata->vmDead) {
+        EXIT_ERROR(AGENT_ERROR_INTERNAL,"VM dead at restore time");
```
Space after comma, please.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1252,6 +1258,46 @@ cbVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+    LOG_CB(("cbVMRestore"));
+
+    BEGIN_CALLBACK() {
+        (void)memset(&info,0,sizeof(info));
```
Space after each comma, please.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -605,6 +623,35 @@ handleReportVMInitCommand(JNIEnv* env, ReportVMInitCommand *command)
+                          JDWP_COMMAND_SET(Event),
+                          JDWP_COMMAND(Event, Composite));
+    (void)outStream_writeByte(&out, suspendPolicy);
+    (void)outStream_writeInt(&out, 1);   /* Always one component */
+    (void)outStream_writeByte(&out, JDWP_EVENT(VM_RESTORE));
+    (void)outStream_writeInt(&out, 0);    /* Not in response to an event req. */
```
Either line up the trailing comments or use just one space before each.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -605,6 +623,35 @@ handleReportVMInitCommand(JNIEnv* env, ReportVMInitCommand *command)
+
+    outStream_sendCommand(&out);
+    outStream_destroy(&out);
+    tossGlobalRef(env, &(command->thread));
```
Please remove the parentheses in `&(command->thread)` (for consistency with e.g. line 685).

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -1167,6 +1219,23 @@ eventHelper_reportVMInit(JNIEnv *env, jbyte sessionID, jthread thread, jbyte sus
+{
+    HelperCommand *command = jvmtiAllocate(sizeof(*command));
+    if (NULL == command) {
+        EXIT_ERROR(AGENT_ERROR_OUT_OF_MEMORY,"HelperCommmand");
+    }
+    memset(command, 0, sizeof(*command));
+    command->commandKind = COMMAND_REPORT_VM_RESTORE;
+    command->sessionID = sessionID;
+    saveGlobalRef(env, thread, &(command->u.reportVMRestore.thread));
```
Space after, comma; extra parentheses.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1918,9 +1923,44 @@ map2jvmtiError(jdwpError error)
+find_ext_event(jvmtiEnv *jvmti, const char *ename)
+{
+    jint extCount = 0;
+    jvmtiExtensionEventInfo* extList = NULL;
```
Space before `*` (here and on line 1927) but not after.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1918,9 +1923,44 @@ map2jvmtiError(jdwpError error)
+            }
+        }
+    } else {
+        ERROR_MESSAGE(("Error in JVMTI GetExtensionFunctions: %s(%d)\n",jvmtiErrorText(err), err));
```
Space after comma.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1918,9 +1923,44 @@ map2jvmtiError(jdwpError error)
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static void
+extensionEventIndexInit(void) {
+    jvmtiExtensionEventInfo* extEvent;
+    /* VM restore event */
+    extEvent = find_ext_event(gdata->jvmti, "VMRestore");
```
Space before, but not after `*`; fold initialization into declaration.

keithc-ca commented on src/java.se/share/data/jdwp/jdwp.spec:
```diff
@@ -3140,6 +3145,17 @@ JDWP "Java(tm) Debug Wire Protocol"
+                        "The event is generated even if not explicitly requested."
+
+                      (int requestID "Request that generated event")
+                        (threadObject thread "The thread restoring the VM from a checkpoint.")
```
Please correct the indentation here.

keithc-ca commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTY.java:
```diff
@@ -99,6 +104,14 @@ public void vmStartEvent(VMStartEvent se)  {
+    @Override
+    public void vmRestoreEvent(VMRestoreEvent se)  {
+        Thread.yield();  // fetch output
+        System.out.print("VM restored from checkpoint: ");
```
Perhaps this should use `MessageOutput` like some other events:
```java
        MessageOutput.lnprint("VM restored from checkpoint:");
```

mikezhang1234567890 replied:
Switched to using `MessageOutput.lnprint`, though that did require me to change another file `TTYResources.java`. Is that fine?

keithc-ca replied:
I think that means you need to change `TTYResources_*.java` as well.

mikezhang1234567890 replied:
Tested it and it pulls from the base resource bundle if it can't find the key in the specific locale bundle.

```
./openj9-openjdk-jdk/build/linux-x86_64-server-release/images/jdk/bin/jdb -attach localhost:8800
nicht abgefangene java.lang.Throwable festlegen
nicht abgefangene java.lang.Throwable mit Verzögerung festlegen
jdb wird initialisiert...
> 
VM restored from checkpoint: "Thread=main", org.eclipse.openj9.criu.CRIUSupport.checkpointJVMImpl(), Zeile=-1 BCI=-1
```

keithc-ca commented on src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java:
```diff
@@ -807,6 +828,12 @@ EventImpl createEvent(JDWP.Event.Composite.Events evt) {
+/*[IF CRIU_SUPPORT]*/
+            case JDWP.EventKind.VM_RESTORE:
+                return new VMRestoreEventImpl(
+                    (JDWP.Event.Composite.Events.VMRestore)comm);
```
Please indent like other cases.

JasonFengJ9 commented on closed/GensrcJ9JCL.gmk:
```diff
@@ -48,6 +48,11 @@ $(eval $(call SetupCopyFiles,COPY_OVERLAY_FILES, \
+		src/java.se/share/data/jdwp/jdwp.spec \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventHandler.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventNotifier.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTY.java \
```
Missed `TTYResources.java`.

mikezhang1234567890 replied:
Added in https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/4e515e9b0139d5148fe5e1599ba67428aa112c0e

mikezhang1234567890 commented:
`j9cfg.h` has been included in every file that uses the `J9VM_OPT_CRIU_SUPPORT` and I believe I addressed the review comments. @keithc-ca can you please take another look?

keithc-ca commented:
I'm not thrilled with this approach. Changing JDWP means no other debuggers will be able to use the new event.
Have you explored reusing `transport_waitForConnection()` at restore time?

mikezhang1234567890 commented:
> Changing JDWP means no other debuggers will be able to use the new event.

Another debugger not using JDWP should still be able to add their own callbacks for the `VMRestore` JVMTI event. The callbacks added to JDWP shouldn't affect that.

> Have you explored reusing `transport_waitForConnection()` at restore time

If a debugger is already attached then `transport_waitForConnection()` will not pause the application.

mikezhang1234567890 commented:
After discussion with @keithc-ca, the issue with adding a new `suspendOnRestore` option is that debuggers are not aware of the new option. It might be better to have the normal `suspend` option apply to starting the VM normally and restoring from a checkpoint, so that people can use their debugging tools as expected on restored applications without making any changes. The new option `suspendOnRestore` could then be used if we want to override this behaviour.

Thoughts on this @tajila ?

tajila commented:
> It might be better to have the normal suspend option apply to starting the VM normally and restoring from a checkpoint

Can you please expand on this part, I dont quite follow. 

Today if I do `-Xrunjdwp:transport=dt_socket,address=*:9900,server=y,suspend=y` it suspends at startup. What are you proposing should happen in checkpoint/restore mode?

mikezhang1234567890 commented:
> > It might be better to have the normal suspend option apply to starting the VM normally and restoring from a checkpoint
> 
> Can you please expand on this part, I dont quite follow.

I mean by default, when `suspend=y`, JDWP would also suspend the VM on restore from checkpoint in addition to suspending on initial VM startup.

So you would launch an application with `java -XX:+EnableCRIUSupport -Xrunjdwp:transport=dt_socket,address=*:9900,server=y,suspend=y App`, and then JDWP would suspend all threads during VM startup. You'd then connect a debugger to resume all threads and the app would take the checkpoint.

When restoring from the checkpointed application, JDWP would then suspend all threads again during the restore process when the JVMTI event is sent.  

If you wanted to not suspend on initial VM startup and only on restore you'd use `-Xrunjdwp:transport=dt_socket,address=*:9900,server=y,suspend=n,suspendOnRestore=y`

tajila commented:
> I mean by default, when suspend=y, JDWP would also suspend the VM on restore from checkpoint in addition to suspending on initial VM startup.

So, this would mean that one would suspend on startup, then attach a debugger to proceed, then checkpoint/restore, then attach a debugger again?

> so that people can use their debugging tools as expected on restored applications without making any changes. 

What changes would be required beyond adding `suspendOnRestore=y`

keithc-ca commented:
From a user's perspective, I think we would prefer _not_ to require any command-line changes.

* If you want to debug prior to the checkpoint, using `suspend=y` should work.
* If you want to debug after the restore, again `suspend=y` is what you should be doing.
* If you want to debug both before and after the checkpoint, you should be able to use `suspend=y` for both launches.

If you don't want to debug a particular phase (or at all) you would launch without the JDWP agent. The VM knows which phase it is executing and should apply the appropriate interpretation to the command-line options.

tajila commented:
With all the new CRIU JVM options we've created we have been very strict about not conflating **startup** with **restore**. There are similarities between the two but it is important to make a distinction as that fundamental understanding is essential to make effective use of checkpoint/restore.

The existing "suspend" option is defined as an option that toggles "wait on startup". From an option specification perspective using suspend to also toggle the default behaviour of suspendOnRestore "wait on restore" adds an inter-dependency between two options which is atypical for option specification.

Personally, Im not sure what the user expectation would be here, so I'm currently asking some of our existing users what their expectation would be.

One thing to note: `suspend=y` is default. Making `suspendOnRestore=y` default as well would mean we debuggers don't need to make changes and it simplifies specification of options.

DanHeidinga commented:
> With all the new CRIU JVM options we've created we have been very strict about not conflating **startup** with **restore**. There are similarities between the two but it is important to make a distinction as that fundamental understanding is essential to make effective use of checkpoint/restore.

+1.  This is key!
> 
> The existing "suspend" option is defined as an option that toggles "wait on startup". From an option specification perspective using suspend to also toggle the default behaviour of suspendOnRestore "wait on restore" adds an inter-dependency between two options which is atypical for option specification.

The docs [0] state: "True if the target VM is to be suspended immediately before the main class is loaded; false otherwise." which doesn't apply at restore as the VM is far past the point of the main class being loaded and there is no equivalent at restore.

[0] https://docs.oracle.com/en/java/javase/11/docs/specs/jpda/conninv.html
> 
> Personally, Im not sure what the user expectation would be here, so I'm currently asking some of our existing users what their expectation would be.
> 
> One thing to note: `suspend=y` is default. Making `suspendOnRestore=y` default as well would mean we debuggers don't need to make changes and it simplifies specification of options.

Making `suspendOnRestore=y` the default makes a lot of sense.  Debuggers won't set it by default but it will be an option if needed.

tajila commented:
> We should aim to make things as simple as possible for those users (certainly not expecting them adjust launch options or worse to build a new debugger).

I think we can get there with `suspendOnRestore=y` as a default setting and not add new meaning to existing options.

mikezhang1234567890 commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -74,6 +80,9 @@ static jboolean initOnUncaught = JNI_FALSE; /* init when uncaught exc thrown */
 static char *launchOnInit = NULL;           /* launch this app during init */
 static jboolean suspendOnInit = JNI_TRUE;   /* suspend all app threads after init */
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static jboolean suspendOnRestore = JNI_TRUE;   /* suspend all app threads after VM restored */
```
Changed default for `suspendOnRestore` to be true.

keithc-ca replied:
Please collapse the repeated spaces:
```c
static jboolean suspendOnRestore = JNI_TRUE; /* suspend all app threads after VM restored */
```

mikezhang1234567890 commented:
Updated to change all uses of `SetEventNotificationMode` to correctly use `SetExtensionEventCallback` for the new `VMRestore` extension event.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1918,9 +1924,43 @@ map2jvmtiError(jdwpError error)
+            }
+        }
+    } else {
+        ERROR_MESSAGE(("Error in JVMTI GetExtensionEvents: %s(%d)\n", jvmtiErrorText(err), err));
```
GetExtensionEvents -> GetExtensionEvent or find_ext_event

mikezhang1234567890 replied:
`GetExtensionEvents` is the correct JVMTI function name.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1918,9 +1924,43 @@ map2jvmtiError(jdwpError error)
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static void
+extensionEventIndexInit(void) {
```
`{` on next line, please.

mikezhang1234567890 commented on closed/GensrcJ9JCL.gmk:
```diff
@@ -50,6 +50,12 @@ $(eval $(call SetupCopyFiles,COPY_OVERLAY_FILES, \
 		src/java.base/share/classes/sun/security/jca/ProviderList.java \
 		src/java.base/unix/classes/java/lang/ProcessEnvironment.java \
+		src/java.se/share/data/jdwp/jdwp.spec \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventHandler.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventNotifier.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTY.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources.java \
+		src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java \
```
All these java files are changed in this PR.

keithc-ca replied:
Sorry, my mistake (I'm not sure what I was comparing when I made that comment).

mikezhang1234567890 commented on src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c:
```diff
@@ -1646,6 +1652,32 @@ threadControl_getInvokeRequest(jthread thread)
+}
+
+static jvmtiExtensionFunction AddDebugThreadToCheckpointState_func = NULL;
+static jvmtiExtensionFunction RemoveDebugThreadFromCheckpointState_func = NULL;
```
Changed to begin with lower case letter, but the function pointer is used at https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591/files#diff-65eb4035b1d9bb5ceaab430461cd782069fe88c0a7d0438964c041caf0aa5eecR1733-R1736 to remove threads from VM storage.

mikezhang1234567890 commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources.java:
```diff
@@ -494,6 +499,9 @@ public Object[][] getContents() {
              "<arguments> are the arguments passed to the main() method of <class>\n" +
              "\n" +
              "For command help type ''help'' at {0} prompt"},
+/*[IF CRIU_SUPPORT]*/
+        {"VM Restored:", "VM restored from checkpoint: "},
+/*[ENDIF] CRIU_SUPPORT */
```
As mentioned in https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591#discussion_r1191714081, if the code can't find the resource for a language/locale, it'll just pull from the base resource.

keithc-ca replied:
I guess I missed that comment, but I don't think that's a very friendly attitude to non-English speakers.

JasonFengJ9 replied:
Updated the non-English versions: `TTYResources_de.java`, `TTYResources_ja.java` and `TTYResources_zh_CN.java`.

mikezhang1234567890 commented:
Do you have any other concerns about this or the openj9 PR @keithc-ca?

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -396,6 +418,30 @@ cbEarlyVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+
+    va_start(ap, jvmti_env);
+    env = va_arg(ap, JNIEnv *);
+    thread = va_arg (ap, jthread);
```
No space before `(`, please.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.h:
```diff
@@ -175,8 +181,13 @@ typedef enum {
+        EI_VM_RESTORE           = 23,
 
+        EI_max                  = 23
+#else
```
This line should be:
```c
#else /* defined(J9VM_OPT_CRIU_SUPPORT) */
```

keithc-ca commented:
I still think the overall behavior should be as I described in https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591#issuecomment-1546248719.

Consider the expanded use-cases where more than one checkpoint is taken in the progression towards increasingly customized applications. The distinction of suspending before a checkpoint versus suspending after one ceases to be clear. Between the first and last checkpoints what would a user be expected to say or do?

I think the answer should be simply, "use the agent, with or without suspend=y, to debug the phase of interest".

tajila commented:
> Consider the expanded use-cases where more than one checkpoint is taken in the progression towards increasingly customized applications. The distinction of suspending before a checkpoint versus suspending after one ceases to be clear.

I think its important to recognize that while in theory there can be infinite checkpoint and restore phases, there can only be one VM start up phase and `suspend` is defined within that context. 

In practical use cases there will only be one restore because of the limitations imposed by security APIs if multiple restores are permitted.

> I think the answer should be simply, "use the agent, with or without suspend=y, to debug the phase of interest".

Im not opposed to that behaviour as a starting point, simply because this feature hasn't been in the field long enough to understand usage patterns. That being said, I can see a need for more flexibility beyond what you suggested. So perhaps something like the following could work: `suspendOnRestore` defaults to `suspend`, this provides the behaviour you suggested and the user can be oblivious to the existence of this option if they are happy with the behaviour. For those who need more control, they can set `suspendOnRestore` to toggle the restore behaviour.

keithc-ca commented:
I have been assuming that debugging settings from one phase are not implicitly carried over into subsequent phases: users must explicitly specify that agent for any phase they want to debug. Is my assumption incorrect or inappropriate?

> I can see a need for more flexibility beyond what you suggested.

Perhaps an example that requires `suspendOnRestore` (i.e. the existing `suspend` option is inadequate or is not appropriate) would help me understand where you see this going.

mikezhang1234567890 commented:
> I have been assuming that debugging settings from one phase are not implicitly carried over into subsequent phases: users must explicitly specify that agent for any phase they want to debug.

The options for the JDWP agent do currently persist through checkpoint and restore.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -605,6 +624,35 @@ handleReportVMInitCommand(JNIEnv* env, ReportVMInitCommand *command)
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static void
+handleReportVMRestoreCommand(JNIEnv* env, ReportVMRestoreCommand *command)
```
Space before, but not after `*` (this line isn't even self-consistent).

mikezhang1234567890 replied:
Fixed in https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/3b8e11025cb78b82767c77f772263c92cd01ad2a

tajila commented:
> I have been assuming that debugging settings from one phase are not implicitly carried over into subsequent phases: users must explicitly specify that agent for any phase they want to debug. Is my assumption incorrect or inappropriate?

JDWP options are only specified at startup. We are unable to dynamically enable JDWP after restore. 

> Perhaps an example that requires suspendOnRestore (i.e. the existing suspend option is inadequate or is not appropriate) would help me understand where you see this going.

An example would be, you want to suspend on start, but not restore.

tajila commented:
> JDWP options are only specified at startup. We are unable to dynamically enable JDWP after restore.

We may have support for this in the future, but this is a med -> long term goal.

keithc-ca commented:
> JDWP options are only specified at startup. We are unable to dynamically enable JDWP after restore.

This seems to me the fundamental issue we should be attacking.

tajila commented:
> I have been assuming that debugging settings from one phase are not implicitly carried over into subsequent phases: users must explicitly specify that agent for any phase they want to debug. Is my assumption incorrect or inappropriate?

Recall, JVM restore is not a new start, it is a continuation of an existing process. So it is correct that options specified on startup should persist on restore. This is true for all JVM options, not just JDWP options. 

We have the ability to specify options on restore. However, if we allowed `suspend` on restore, we would be redefining the meaning of suspend, because suspend is defined in the context of VM startup. This is the fundamental issue with what you are proposing. 

However, if we had `suspendOnRestore` it is understood that this option is defined in the context of restore. For example:

Phases -> effect
//startup options: suspend=y,suspendOnRestore=y
Startup -> VM suspends
1st Restore -> VM suspends
//restore options: suspendOnRestore=n
2nd Restore -> VM does not suspend
//restore options: suspendOnRestore=y
3rd Restore -> VM suspends

> > JDWP options are only specified at startup. We are unable to dynamically enable JDWP after restore.

> This seems to me the fundamental issue we should be attacking.

When we are able to dynamically enable debug capabilities, we will be able to do:

Phases -> effect
Startup -> VM does not suspend
//restore options: suspendOnRestore=y
1st Restore -> VM suspends
//restore options: suspendOnRestore=n
2nd Restore -> VM does not suspend
//restore options: suspendOnRestore=y
3rd Restore -> VM suspends

So I dont see it as an issue, it is really a point in time limitation. 

One could make the arguement, that `suspendOnRestore` should only be a restore option, that is, you cant specify it on startup but you can on restore. I wouldn't be opposed to that. But I think there needs to be a clear distiction between, suspend on VM start, and suspend on restore.

keithc-ca commented:
I repeat my concerns about this approach.

In order to debug *any* phase in the evolution of an application through one or more checkpoints, the JDWP agent must be enabled in the very first phase (and based on comments above, every phase). The producer of an application *must* enable the agent unless they are prepared to accept the consequence that no customers will be able to debug their use of the product.

Enabling the JDWP agent has associated performance implications: all subsequent phases will pay those costs even when debugging is not needed. Using `suspendOnRestore=y` (or making it the default) would mean that a debugger must be used to attach to the application to instruct it to proceed - not a very user-friendly situation.

It may be that the VM needs to provide suspend and restore events, but to my mind this doesn't represent the way they should be handled.

mikezhang1234567890 commented:
> the JDWP agent must be enabled in the very first phase (and based on comments above, every phase)

Currently, even without CRIU, the JDWP agent must be enabled at start up due to the capabilities needed. Regardless of CRIU, using JDWP for debug means you're taking the performance hit for the entire run of an application.

> The producer of an application must enable the agent unless they are prepared to accept the consequence that no customers will be able to debug their use of the product.

Enabling JDWP is the user's decision at run time, so consequences are determined by the user each run.

JasonFengJ9 commented:
Added a commit [CRIU supports Java debugger during checkpoint and restore](https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591/commits/815a9f4709f35f1f73571aa12027ff3cc5ac85ee)
* Added an option suspendOnRestore which suspends when the VM is restored
from a checkpoint;
* The existing option "suspend=y" also suspends the VM at CRIU restore.

JasonFengJ9 commented:
@tajila @keithc-ca could you please review this PR to enable [CRIU supports Java debugger during checkpoint and restore](https://github.com/eclipse-openj9/openj9/pull/19405)?

Note: I added a commit [CRIU supports Java debugger during checkpoint and restore](https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591/commits/815a9f4709f35f1f73571aa12027ff3cc5ac85ee) at the top of initial changes.

JasonFengJ9 commented:
JDK21 benchmark results with this PR (default settings, no JDWP):

* [jdk21_20240529_152441 - baseline](https://na-public.artifactory.swg-devops.com/artifactory/sys-rt-generic-local/hyc-runtimes-jenkins.swg-devops.com/Build_JDK21_x86-64_linux_Personal/654/OpenJ9-JDK21-x86-64_linux-20240529-152441.tar.gz) 
 openj9 repo - `master` branch w/o https://github.com/eclipse-openj9/openj9/pull/19405
 extension repo - `openj9` branch

```
Startup time results for localhost/ol-instanton-test-pingperf-restore-afterappstart:jdk21_20240529_152441 app pingperf restoreOpts None
```
| Bechmark metrics | avg | min | max | stdDev | maxVar | confInt | samples |
| ------------------ | ---- | ----| ---- | ------- | -------- | ------- | -------- |
| StartupTime | 531 | 503 | 569 | 13 | 13.1% | 0.47% | 80 |
| CRIUTime | 87 | 87 | 88 | 0 | 1.8% | 0.06% | 68 |
| Application | 153 | 148 | 157 | 2 | 6.1% | 0.26% | 80 |
| FirstResponse | 632 | 600 | 672 | 14 | 12.0% | 0.42% | 80 |
| ContainerFootprint (KiB) | 117617 | 117043 | 118067 | 233 | 0.9% | 0.04% | 80 |
| RSS (KiB) | 149039 | 148828 | 149244 | 102 | 0.3% | 0.01% | 80 |
| Peak RSS (KiB) | 147954 | 146064 | 151600 | 942 | 3.8% | 0.12% | 80 |

```
Throughput results for localhost/ol-instanton-test-pingperf-restore-afterappstart:jdk21_20240529_152441 app pingperf restoreOpts None
```
| Bechmark metrics | avg | min | max | stdDev | maxVar | confInt | samples |
| ------------------ | ---- | ----| ---- | ------- | -------- | ------- | -------- |
| Throughput | 23677 | 23118 | 24147 | 337 | 4.5% | 0.82% | 10 |
| ContainerFootprint (KiB) | 896266 | 868966 | 914432 | 13487 | 5.2% | 0.87% | 10 |
| RSS (KiB) | 186825 | 181600 | 193412 | 3930 | 6.5% | 1.22% | 10 |
| Peak RSS (KiB) | 428391 | 415568 | 458324 | 11971 | 10.3% | 1.62% | 10 |

* [jdk21_20240529_152833 - C/R debugger support via startup command line option](https://na-public.artifactory.swg-devops.com/artifactory/sys-rt-generic-local/hyc-runtimes-jenkins.swg-devops.com/Build_JDK21_x86-64_linux_Personal/656/OpenJ9-JDK21-x86-64_linux-20240529-152833.tar.gz)
  openj9 repo - `master` branch containing https://github.com/eclipse-openj9/openj9/pull/19405
  extension repo - `openj9` branch ported this PR


```
Startup time results for localhost/ol-instanton-test-pingperf-restore-afterappstart:jdk21_20240529_152833 app pingperf restoreOpts None
```
| Bechmark metrics | avg | min | max | stdDev | maxVar | confInt | samples |
| ------------------ | ---- | ----| ---- | ------- | -------- | ------- | -------- |
| StartupTime | 533 | 503 | 558 | 13 | 10.9% | 0.47% | 80 |
| CRIUTime | 88 | 88 | 89 | 0 | 1.5% | 0.06% | 67 |
| Application | 152 | 151 | 155 | 1 | 2.6% | 0.10% | 80 |
| FirstResponse | 634 | 603 | 658 | 14 | 9.1% | 0.40% | 80 |
| ContainerFootprint (KiB) | 114707 | 114176 | 115098 | 189 | 0.8% | 0.03% | 79 |
| RSS (KiB) | 146616 | 146336 | 147072 | 132 | 0.5% | 0.02% | 80 |
| Peak RSS (KiB) | 145651 | 143568 | 147788 | 936 | 2.9% | 0.12% | 80 |


```
Throughput results for localhost/ol-instanton-test-pingperf-restore-afterappstart:jdk21_20240529_152833 app pingperf restoreOpts None
```
| Bechmark metrics | avg | min | max | stdDev | maxVar | confInt | samples |
| ------------------ | ---- | ----| ---- | ------- | -------- | ------- | -------- |
| Throughput | 23706 | 22859 | 24058 | 362 | 5.2% | 0.89% | 10 |
| ContainerFootprint (KiB) | 900045 | 869888 | 912282 | 12294 | 4.9% | 0.79% | 10 |
| RSS (KiB) | 182587 | 179436 | 187756 | 2801 | 4.6% | 0.89% | 10 |
| Peak RSS (KiB) | 430518 | 403116 | 453400 | 18868 | 12.5% | 2.94% | 8 |

This benchmark result shows that this PR has no noticeable perf impact when the JDWP is not enabled.

JasonFengJ9 commented:
jenkins copyright check

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c:
```diff
@@ -1669,7 +1701,14 @@ threadControl_addDebugThread(jthread thread)
             error = AGENT_ERROR_OUT_OF_MEMORY;
         } else {
             debugThreadCount++;
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+            if (NULL == addDebugThreadToCheckpointState_func) {
+                addDebugThreadToCheckpointState_func = find_ext_function(gdata->jvmti, "AddDebugThreadToCheckpointState");
+            }
+            error = (*addDebugThreadToCheckpointState_func)(gdata->jvmti, thread);
+#else /* defined(J9VM_OPT_CRIU_SUPPORT) */
             error = JVMTI_ERROR_NONE;
+#endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
```
Suggest the `#else` code is not needed:
```c
            error = JVMTI_ERROR_NONE;
#if defined(J9VM_OPT_CRIU_SUPPORT)
            if (NULL == addDebugThreadToCheckpointState_func) {
                addDebugThreadToCheckpointState_func = find_ext_function(gdata->jvmti, OPENJ9_FUNCTION_ADD_DEBUG_THREAD);
            }
            error = (*addDebugThreadToCheckpointState_func)(gdata->jvmti, thread);
#endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
```

JasonFengJ9 replied:
Removed `#else`.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c:
```diff
@@ -1696,7 +1735,14 @@ threadControl_removeDebugThread(jthread thread)
                 debugThreads[j-1] = debugThreads[j];
             }
             debugThreadCount--;
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+            if (NULL == removeDebugThreadFromCheckpointState_func) {
+                removeDebugThreadFromCheckpointState_func = find_ext_function(gdata->jvmti, "RemoveDebugThreadFromCheckpointState");
+            }
+            error = (*removeDebugThreadFromCheckpointState_func)(gdata->jvmti, thread);
+#else /* defined(J9VM_OPT_CRIU_SUPPORT) */
             error = JVMTI_ERROR_NONE;
+#endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
```
Suggest the `#else` code is not needed:
```c
            error = JVMTI_ERROR_NONE;
#if defined(J9VM_OPT_CRIU_SUPPORT)
            if (NULL == removeDebugThreadFromCheckpointState_func) {
                removeDebugThreadFromCheckpointState_func = find_ext_function(gdata->jvmti, OPENJ9_FUNCTION_REMOVE_DEBUG_THREAD);
            }
            error = (*removeDebugThreadFromCheckpointState_func)(gdata->jvmti, thread);
#endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
```

JasonFengJ9 replied:
Removed `#else`.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/transport.c:
```diff
@@ -345,6 +352,24 @@ transport_waitForConnection(void)
+     * If suspended on VM restore, we also need to wait for a connection,
+     * since the VM won't continue without a remote debugger telling it to.
+     */
+    if (debugInit_suspendOnRestore()) {
```
Suggestions elsewhere make this call to `debugInit_suspendOnRestore()` redundant; suggest it be removed.

JasonFengJ9 replied:
Removed this `debugInit_suspendOnRestore())` call.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1901,9 +1908,44 @@ map2jvmtiError(jdwpError error)
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static jvmtiExtensionEventInfo *
+find_ext_event(jvmtiEnv *jvmti, const char *ename)
+{
+    jint extCount = 0;
+    jvmtiExtensionEventInfo *extList = NULL;
+
+    jvmtiError err = JVMTI_FUNC_PTR(jvmti, GetExtensionEvents)
+                            (jvmti, &extCount, &extList);
+    if (JVMTI_ERROR_NONE == err) {
+        for (int i = 0; i < extCount; i++) {
+            if (NULL != strstr(extList[i].id, ename)) {
```
A better name for `ename` is `eventName`.
This should use `strcmp()`.
In C code, locals cannot be declared in `for` loops.
```c
find_ext_event(jvmtiEnv *jvmti, const char *eventName)
{
    jint extCount = 0;
    jvmtiExtensionEventInfo *extList = NULL;

    jvmtiError err = JVMTI_FUNC_PTR(jvmti, GetExtensionEvents)
                            (jvmti, &extCount, &extList);
    if (JVMTI_ERROR_NONE == err) {
        jint i = 0;
        for (i = 0; i < extCount; i++) {
            if (0 == strcmp(extList[i].id, eventName)) {
```

JasonFengJ9 replied:
Updated, and also avoid multiple returns.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1901,9 +1908,44 @@ map2jvmtiError(jdwpError error)
 static jvmtiEvent index2jvmti[EI_max-EI_min+1];
 static jdwpEvent  index2jdwp [EI_max-EI_min+1];
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
```
This will be helpful with the update to `find_ext_event()`
```c
#define OPENJ9_EVENT_VM_RESTORE "openj9.event.VMRestore"
```

JasonFengJ9 replied:
Use OpenJ9 define.

JasonFengJ9 commented on closed/src/jdk.jdi/share/classes/com/sun/jdi/event/VMRestoreEvent.java:
```diff
@@ -0,0 +1,50 @@
+/*[INCLUDE-IF CRIU_SUPPORT]*/
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2024, 2024 All Rights Reserved
```
Updated

JasonFengJ9 commented on closed/GensrcJ9JCL.gmk:
```diff
@@ -58,6 +58,12 @@ $(eval $(call SetupCopyFiles,COPY_OVERLAY_FILES, \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventHandler.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventNotifier.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTY.java \
+		src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources.java \
```
Added

JasonFengJ9 commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/EventHandler.java:
```diff
@@ -151,6 +156,10 @@ private boolean handleEvent(Event event) {
             return vmStartEvent(event);
+/*[IF CRIU_SUPPORT]*/
+        } else if (event instanceof VMRestoreEvent) {
+            return vmRestoreEvent(event);
```
Updated

JasonFengJ9 commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTY.java:
```diff
@@ -99,6 +105,14 @@ public void vmStartEvent(VMStartEvent se)  {
+/*[IF CRIU_SUPPORT]*/
+    @Override
+    public void vmRestoreEvent(VMRestoreEvent re)  {
+        Thread.yield();  // fetch output
```
Removed the extra space.

JasonFengJ9 commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources_de.java:
```diff
@@ -343,6 +349,9 @@ public Object[][] getContents() {
         {"zz usage text",
              "Verwendung: {0} <Optionen> <Klasse> <Argumente>\n\nVerf\u00FCgbare Optionen:\n    -? -h --help -help Gibt diese Hilfemeldung aus und beendet den Vorgang\n    -sourcepath <durch \"{1}\" getrennte Verzeichnisse>\n                      Verzeichnisse, in denen nach Quelldateien gesucht werden soll\n    -attach <Adresse>\n                      Anh\u00E4ngen an aktive VM unter der angegeb ...
+/*[IF CRIU_SUPPORT]*/
+        {"VM Restored:", "\u0056\u004d\u0020\u0076\u006f\u006d\u0020\u0050\u0072\u00fc\u0066\u0070\u0075\u006e\u006b\u0074\u0020\u0077\u0069\u0065\u0064\u0065\u0072\u0068\u0065\u0072\u0067\u0065\u0073\u0074\u0065\u006c\u006c\u0074\u003a"},
```
Changed `TTYResources_de.java` to avoid unicode, and also changed the unicode to uppercase hex digits.

JasonFengJ9 commented on src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java:
```diff
@@ -64,6 +70,9 @@
 import com.sun.jdi.event.VMDeathEvent;
 import com.sun.jdi.event.VMDisconnectEvent;
 import com.sun.jdi.event.VMStartEvent;
+/*[IF CRIU_SUPPORT]*/
+import com.sun.jdi.event.VMRestoreEvent;
+/*[ENDIF] CRIU_SUPPORT */
```
Moved it up.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -330,6 +343,16 @@ DEF_Agent_OnLoad(JavaVM *vm, char *options, void *reserved)
                         jvmtiErrorText(error), error));
         return JNI_ERR;
     }
+    /* enable extension events and set early callbacks for them */
```
Removed repeated spaces.
Moved the comment down.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -711,6 +767,17 @@ initialize(JNIEnv *env, jthread thread, EventIndex triggering_ei, EventInfo *opt
+
+        LOG_MISC(("triggering_ei == EI_VM_RESTORE"));
+        suspendPolicy = suspendOnRestore ? JDWP_SUSPEND_POLICY(ALL) : JDWP_SUSPEND_POLICY(NONE);
+        eventHelper_reportVMRestore(env, currentSessionID, thread, suspendPolicy);
```
Updated with conditional check.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -1198,6 +1273,15 @@ parseOptions(char *options)
                 goto syntax_error;
             }
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+            if (suspendOnInit) {
+                suspendOnRestore = JNI_TRUE;
+            }
```
> Should this override suspendOnRestore=false specified earlier?

Not sure about the reference of `suspendOnRestore=false`.
`suspendOnRestore` is initialized to `JNI_TRUE` to indicate that by default the VM will be suspended post-restore if a `jdwp` agent is specified pre-checkpoint such as `-agentlib:jdwp=transport=dt_socket,server=y,suspend=n`
If `suspendOnRestore=n` is also supplied in that command line option `-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,suspendOnRestore=n`, the VM won't be suspended post-restore.

keithc-ca replied:
Sorry, I expected boolean options to accept "true" or "false": I should have asked about the interaction with `suspendOnRestore=n`.

To make it more concrete, `-agentlib:jdwp=transport=dt_socket,server=y,suspendOnRestore=n,suspend=y` will be handled as if the user said `-agentlib:jdwp=transport=dt_socket,server=y,suspendOnRestore=y,suspend=y` which may be surprising.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -68,6 +75,8 @@
 #include "debugLoop.h"
 #include "signature.h"
+#include "transport.h"
+#include "j9cfg.h"
```
Added `J9VM_OPT_CRIU_SUPPORT` decoration.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1252,6 +1261,46 @@ cbVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+    EventInfo info;
+    JNIEnv *env = NULL;
+    jthread thread = NULL;
+    va_list ap;
```
Replaced it with `args`.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1252,6 +1261,46 @@ cbVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+    JNIEnv *env = NULL;
+    jthread thread = NULL;
+    va_list ap;
+    jbyte suspendPolicy = 0;
```
Updated

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1252,6 +1261,46 @@ cbVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+    BEGIN_CALLBACK() {
+        (void)memset(&info, 0, sizeof(info));
+        info.ei         = EI_VM_RESTORE;
+        info.thread     = thread;
```
Removed those spaces.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1252,6 +1261,46 @@ cbVMInit(jvmtiEnv *jvmti_env, JNIEnv *env, jthread thread)
+     * event handler code similar to the VM init event
+     */
+    suspendPolicy = debugInit_suspendOnRestore() ? JDWP_SUSPEND_POLICY(ALL) : JDWP_SUSPEND_POLICY(NONE);
+    eventHelper_reportVMRestore(env, currentSessionID, thread, suspendPolicy);
```
Updated the comments, and add the conditional check.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:
```diff
@@ -1273,6 +1322,15 @@ cbVMDeath(jvmtiEnv *jvmti_env, JNIEnv *env)
     }
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+    /* clear extension event callbacks */
```
Updated

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -90,6 +100,13 @@ typedef struct ReportVMInitCommand {
+    jbyte suspendPolicy; /* NOTE: Must be the first field */
+    jthread thread;
+} ReportVMRestoreCommand;
+#endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
```
Removed `ReportVMRestoreCommand`, and reuse `ReportVMInitCommand` instead.

keithc-ca replied:
Part of the motivation for not defining a new type was to avoid adding a new member to the `union` in `HelperCommand`. Please just use `reportVMInit` instead of defining `reportVMRestore`.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -605,6 +625,35 @@ handleReportVMInitCommand(JNIEnv* env, ReportVMInitCommand *command)
+    (void)outStream_writeByte(&out, suspendPolicy);
+    (void)outStream_writeInt(&out, 1); /* Always one component */
+    (void)outStream_writeByte(&out, JDWP_EVENT(VM_RESTORE));
+    (void)outStream_writeInt(&out, 0); /* Not in response to an event req. */
```
Updated

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c:
```diff
@@ -1651,6 +1658,31 @@ threadControl_getInvokeRequest(jthread thread)
+    if (JVMTI_ERROR_NONE == err) {
+        for (int i = 0; i < extCount; i++) {
+            if (NULL != strstr(extList[i].id, fname)) {
+                return extList[i].func;
```
Updated, and also avoid multiple returns.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/threadControl.c:
```diff
@@ -1651,6 +1658,31 @@ threadControl_getInvokeRequest(jthread thread)
+    return NULL;
+}
+
+static jvmtiExtensionFunction addDebugThreadToCheckpointState_func = NULL;
```
Use the OpenJ9 defines.

keithc-ca commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTY.java:
```diff
@@ -99,6 +105,14 @@ public void vmStartEvent(VMStartEvent se)  {
 
+/*[IF CRIU_SUPPORT]*/
+    @Override
+    public void vmRestoreEvent(VMRestoreEvent re)  {
```
Please collapse the double-space before `{`.

keithc-ca commented on src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources_de.java:
```diff
@@ -343,6 +349,9 @@ public Object[][] getContents() {
         {"zz usage text",
              "Verwendung: {0} <Optionen> <Klasse> <Argumente>\n\nVerf\u00FCgbare Optionen:\n    -? -h --help -help Gibt diese Hilfemeldung aus und beendet den Vorgang\n    -sourcepath <durch \"{1}\" getrennte Verzeichnisse>\n                      Verzeichnisse, in denen nach Quelldateien gesucht werden soll\n    -attach <Adresse>\n                      Anh\u00E4ngen an aktive VM unter der angegeb ...
+/*[IF CRIU_SUPPORT]*/
+        {"VM Restored:", "VM vom Prüfpunkt wiederhergestellt:"},
```
I think `ü` should be written `\u00FC` like on line 351.

keithc-ca commented on src/jdk.jdi/share/classes/com/sun/tools/jdi/EventSetImpl.java:
```diff
@@ -496,6 +505,19 @@ String eventName() {
+            super(evt, evt.requestID, evt.thread);
+        }
+
+        String eventName() {
```
This is missing the `@Override` annotation.

JasonFengJ9 replied:
Added

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -101,6 +111,9 @@ static void JNICALL cbEarlyVMInit(jvmtiEnv*, JNIEnv *, jthread);
 static void JNICALL cbEarlyException(jvmtiEnv*, JNIEnv *,
             jthread, jmethodID, jlocation, jobject, jmethodID, jlocation);
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static void JNICALL cbEarlyVMRestore(jvmtiEnv*, ...);
```
Space before `*`, please.

JasonFengJ9 replied:
Added a space.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -34,7 +40,10 @@
 #include "inStream.h"
 #include "invoker.h"
 #include "signature.h"
-
```
Please leave this line from upstream alone.

JasonFengJ9 replied:
Restore the empty line.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/util.c:
```diff
@@ -1901,9 +1910,47 @@ map2jvmtiError(jdwpError error)
 }
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+static jvmtiExtensionEventInfo*
```
Space before `*`, please.

JasonFengJ9 replied:
Added a space.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -594,15 +604,20 @@ handleReportVMInitCommand(JNIEnv* env, ReportVMInitCommand *command)
-    (void)outStream_writeInt(&out, 1);   /* Always one component */
-    (void)outStream_writeByte(&out, JDWP_EVENT(VM_INIT));
-    (void)outStream_writeInt(&out, 0);    /* Not in response to an event req. */
+    (void)outStream_writeInt(&out, 1); /* Always one component. */
+    (void)outStream_writeByte(&out, jdwpEvent);
+    (void)outStream_writeInt(&out, 0); /* Not in response to an event request. */
```
Please leave lines 607 and 609 as is.

JasonFengJ9 replied:
Restored the upstream comments.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/eventHelper.c:
```diff
@@ -594,15 +604,20 @@ handleReportVMInitCommand(JNIEnv* env, ReportVMInitCommand *command)
     /* Why aren't we tossing this: tossGlobalRef(env, &(command->thread)); */
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+    if (JDWP_EVENT(VM_RESTORE) == jdwpEvent) {
+        tossGlobalRef(env, &command->thread);
```
Please indent only with spaces.

JasonFengJ9 replied:
Yeah, it failed the Line Endings Check initially and was fixed afterwards.

JasonFengJ9 commented:
> It may be reasonable that suspend=y alters the default for suspendOnRestore, but it should not override an explicit specification.

Changed the command line option precedence as follows:
* `-agentlib:jdwp=transport=dt_socket,server=y,suspend=[y|n]` assumes `suspendOnRestore=y`;
* Always respect `suspendOnRestore=n` regardless the order of `suspend` and `suspendOnRestore`, both `-agentlib:jdwp=transport=dt_socket,server=y,suspendOnRestore=n,suspend=y` and `-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,suspendOnRestore=n` don't suspend the JVM post-restore.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -74,6 +81,14 @@ static jboolean initOnUncaught = JNI_FALSE; /* init when uncaught exc thrown */
 static char *launchOnInit = NULL;           /* launch this app during init */
 static jboolean suspendOnInit = JNI_TRUE;   /* suspend all app threads after init */
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+/* By default VM is suspended post-restore unless suspendOnRestore=n is specified.
+ * -1 initial value
+ *  0 not suspend post-restore
+ *  1 suspend post-restore
+ */
+static jbyte suspendOnRestore = -1;
```
Suggest this continue to have type `jboolean` (else the use of `debugInit_suspendOnRestore()` is incorrect).
A local (boolean) variable in `parseOptions()` can be used to help distinguish these cases.

JasonFengJ9 replied:
Added a local `jboolean suspendOnRestoreTmp = JNI_TRUE`, this is to be updated according to `suspendOnRestore=[y|n]`, and assigned to the global `suspendOnRestore`.

keithc-ca commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -1207,6 +1286,14 @@ parseOptions(char *options)
                 goto syntax_error;
             }
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+            suspendOnRestore = suspendOnRestoreTmp;
```
I don't think this does what we want for `suspend=n` (which should imply a default of `suspendOnRestore=n`).

JasonFengJ9 replied:
This was proposed at https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591#issuecomment-2296411187
* -agentlib:jdwp=transport=dt_socket,server=y,suspend=[y|n] assumes suspendOnRestore=y

With `-XX:+DebugOnRestore -XX:+EnableCRIUSupport`, the assumption is the user would like to debug post-restore, particularly the post-restore hook application code. Only with `suspendOnRestore=y`, the user has the chance to add breakpoints in those hooks.
If `suspendOnRestore=n` is specified explicitly, it will be honoured and JVM won't be suspended post-restore.

keithc-ca replied:
I'm not sure that's the behavior we ultimately want, but for now, let's assume it is.

I don't think `suspendOnRestoreTmp` is needed. The assigment on line 1292 does nothing (if `suspendOnRestoreTmp` is true (either because it's the default or the user specified `suspendOnRestore=y`), then `suspendOnRestore` is already true; if `suspendOnRestoreTmp` is false, it's because the user said `suspendOnRestore=n` in which case `suspendOnRestore` is already false. Line 1292 can simply be deleted, and line 1294 should use `suspendOnRestore`.

JasonFengJ9 replied:
Yeah, the temp isn't needed, removed it.

JasonFengJ9 commented on src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:
```diff
@@ -984,6 +1062,7 @@ parseOptions(char *options)
     char *str;
     char *errmsg;
     jboolean onJcmd = JNI_FALSE;
+    jboolean suspendOnRestoreTmp = JNI_TRUE;
```
Added `J9VM_OPT_CRIU_SUPPORT` define.

keithc-ca commented on closed/GensrcJ9JCL.gmk:
```diff
@@ -57,7 +57,16 @@ $(eval $(call SetupCopyFiles,COPY_OVERLAY_FILES, \
+<<<<<<< Upstream, based on openj9
 		src/java.base/unix/classes/sun/security/provider/NativePRNG.java \
+=======
+		src/java.se/share/data/jdwp/jdwp.spec \
```
Please squash so we don't have commits with unresolved merge issues like this.

JasonFengJ9 replied:
Squashed to remove unresolved merge conflict.

JasonFengJ9 commented:
This PR depended on https://github.com/eclipse-openj9/openj9/pull/19405 which has been merged.

* https://github.com/eclipse-openj9/openj9/pull/19754
depends on this PR instead.

keithc-ca commented:
Jenkins copyright check

AdamBrousseau commented:
jenkins copyright check

AdamBrousseau commented:
jenkins copyright check

AdamBrousseau commented:
jenkins copyright check

AdamBrousseau commented:
jenkins copyright check

AdamBrousseau commented:
jenkins copyright check

keithc-ca commented:
Jenkins test sanity amac,xlinux,zlinux jdknext

JasonFengJ9 commented:
> 17:16:43  {standard input}: Fatal error: /home/jenkins/workspace/Build_JDKnext_x86-64_linux_Personal/build/linux-x86_64-server-release/support/test/hotspot/jtreg/native/support/libsetjniftab001/libsetjniftab001.o: No space left on device

Infra issue

tajila commented:
jenkins test sanity xlinux jdknext

keithc-ca commented:
@tajila If you're happy with this, could you please formally approve?

JasonFengJ9 commented:
Backports:

* https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/803
* https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/377
* https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/189
* https://github.com/ibmruntimes/openj9-openjdk-jdk23/pull/34

JDK22 is out of service, JDK8 doesn't support CRIU.
